File: Appstax/Appstax/AXLoginViewController.m

#import "AXLoginViewController.h"
#import "Appstax.h"
#import "AppstaxInternals.h"
#import "AXUserService.h"

@interface AXLoginViewController ()
@property (weak, nonatomic) IBOutlet UITextField *usernameTextField;
@property (weak, nonatomic) IBOutlet UITextField *passwordTextField;
@property (weak, nonatomic) IBOutlet UIButton *submitButton;
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *activityIndicator;
@property (weak, nonatomic) IBOutlet UILabel *errorLabel;
@property (weak) AXLoginUIManager *loginUIManager;
@property (weak, nonatomic) IBOutlet UIButton *goToLoginButton;
@property (weak, nonatomic) IBOutlet UIView *backgroundContainer;
@property (nonatomic) UIView *backgroundView;
@property BOOL viewLoaded;
@end

@implementation AXLoginViewController


- (instancetype)initWithUIManager:(AXLoginUIManager *)uiManager {
    self = [super initWithNibName:@"AXLoginViewController" bundle:[Appstax frameworkBundle]];
    if (self) {
        _loginUIManager = uiManager;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    _viewLoaded = YES;
    [_submitButton setTitle:_submitTitle forState:UIControlStateNormal];
    _goToLoginButton.hidden = _goToLoginHidden;
    [self setBackgroundView:_backgroundView];
}

-(BOOL)textFieldShouldReturn:(UITextField *)textField {
    if (textField == _usernameTextField) {
        [_passwordTextField becomeFirstResponder];
    } else if(textField == _passwordTextField) {
        [textField resignFirstResponder];
        [self pressedSignUpButton:nil];
    }
    return YES;
}

- (IBAction)pressedSignUpButton:(id)sender {
    [self disableInputs];
    [_activityIndicator startAnimating];
    _errorLabel.hidden = YES;
    [_loginUIManager viewControllerDidPressSubmitButton:self];
}

- (IBAction)pressedGoToLoginButton:(id)sender {
    [_loginUIManager viewControllerDidPressGoToLoginButton:self];
}

- (void)setBackgroundView:(UIView *)view {
    if(_backgroundView != nil) {
        [_backgroundView removeFromSuperview];
        _backgroundView = nil;
    }
    _backgroundView = view;
    if(_viewLoaded) {
        _backgroundView.frame = _backgroundContainer.bounds;
        [_backgroundContainer addSubview:_backgroundView];
    }
}

- (void)clear {
    _usernameTextField.text = @"";
    _passwordTextField.text = @"";
    _errorLabel.text = @"";
    [_activityIndicator stopAnimating];
    [self enableInputs];
}

- (void)showError:(NSString *)errorMessage {
    _errorLabel.hidden = NO;
    _errorLabel.text = errorMessage;
    [self enableInputs];
    [_activityIndicator stopAnimating];
}

- (void)disableInputs {
    [self.view endEditing:YES];
    _usernameTextField.enabled = NO;
    _passwordTextField.enabled = NO;
    _submitButton.enabled = NO;
}

- (void)enableInputs {
    _usernameTextField.enabled = YES;
    _passwordTextField.enabled = YES;
    _submitButton.enabled = YES;
}

- (NSString *)username {
    return _usernameTextField.text;
}

- (NSString *)password {
    return _passwordTextField.text;
}

@end

File: Appstax/Appstax/AXLoginViewController.swift

import UIKit

class AXLoginViewController: UIViewController {
    
    
    @IBOutlet var facebookButton: UIButton!
    @IBOutlet var googleButton: UIButton!
    
    @IBOutlet weak var usernameTextField: UITextField!
    @IBOutlet weak var passwordTextField: UITextField!
    @IBOutlet weak var submitButton: UIButton!
    @IBOutlet weak var activityIndicator: UIActivityIndicatorView!
    @IBOutlet weak var errorLabel: UILabel!
    @IBOutlet weak var goToLoginButton: UIButton!
    @IBOutlet weak var backgroundContainer: UIView!
    @IBOutlet weak var providerButtonsContainer: UIView!
    @IBOutlet weak var providerButtonsHeightConstraint: NSLayoutConstraint!
    
    var submitTitle: String = ""
    var goToLoginHidden: Bool = false
    
    private var viewLoaded = false
    private var loginUIManager: AXLoginUIManager!
    
    var backgroundView: UIView? {
        willSet {
            backgroundView?.removeFromSuperview()
        }
        didSet {
            ensureViewLoaded()
            if let bg = backgroundView {
                bg.frame = backgroundContainer.bounds
                backgroundContainer.addSubview(bg)
            }
        }
    }
    
    var providers: [String] = [] {
        didSet {
            updateProviderButtons();
        }
    }
    
    var username: String {
        get {
            return usernameTextField.text ?? ""
        }
    }
    var password: String {
        get {
            return passwordTextField.text ?? ""
        }
    }
    
    convenience init(manager: AXLoginUIManager) {
        self.init(nibName: "AXLoginViewController", bundle:Appstax.frameworkBundle())
        self.loginUIManager = manager
    }
    
    override func viewDidLoad() {
        viewLoaded = true
        submitButton.setTitle(submitTitle, forState: .Normal)
        goToLoginButton.hidden = goToLoginHidden
    }
    
    func ensureViewLoaded() {
        _ = self.view
    }
    
    func textFieldShouldReturn(textField: UITextField) -> Bool {
        if textField === usernameTextField {
            passwordTextField.becomeFirstResponder()
        } else if textField == passwordTextField {
            textField.resignFirstResponder()
            pressedSignUpButton(nil)
        }
        return true
    }
    
    @IBAction func pressedSignUpButton(sender: AnyObject?) {
        disableInputs()
        activityIndicator.startAnimating()
        errorLabel.hidden = true
        loginUIManager.viewControllerDidPressSubmitButton(self)
    }
    
    @IBAction func pressedGoToLoginButton(sender: AnyObject?) {
        loginUIManager.viewControllerDidPressGoToLoginButton(self)
    }
    
    func updateProviderButtons() {
        let buttonHeight: CGFloat = 44
        let buttonSpacing: CGFloat = 10
        let buttonWidth = providerButtonsContainer.frame.width
        
        var y = CGFloat(0)
        
        providerButtonsContainer.subviews.forEach({ v in v.removeFromSuperview() })
        providers.enumerate().forEach({
            index, provider in
            
            if let button = buttonForProvider(provider) {
                button.tag = index
                button.frame = CGRect(x: 0, y: y, width: buttonWidth, height: buttonHeight)
                button.removeTarget(nil, action: nil, forControlEvents: .AllEvents)
                button.addTarget(self, action: #selector(AXLoginViewController.didPressProviderButton(_:)), forControlEvents: .TouchUpInside)
                
                providerButtonsContainer.addSubview(button)
                y += buttonHeight + buttonSpacing
            }
        })
        
        providerButtonsHeightConstraint.constant = y
    }
    
    func buttonForProvider(provider: String) -> UIButton? {
        switch(provider) {
            case "facebook": return facebookButton
            case "google":   return googleButton
            default: return nil
        }
    }
    
    func didPressProviderButton(button: UIButton) {
        if providers.count > button.tag {
            let provider = providers[button.tag]
            loginUIManager.viewControllerDidPressProviderButton(self, provider: provider)
        }
    }
    
    func clear() {
        usernameTextField.text = ""
        passwordTextField.text = ""
        errorLabel.text = ""
        activityIndicator.stopAnimating()
        enableInputs()
    }
    
    func showError(errorMessage: String) {
        errorLabel.hidden = false
        errorLabel.text = errorMessage
        enableInputs()
        activityIndicator.stopAnimating()
    }
    
    func disableInputs() {
        self.view.endEditing(true)
        usernameTextField.enabled = false
        passwordTextField.enabled = false
        submitButton.enabled = false
    }
    
    func enableInputs() {
        usernameTextField.enabled = true
        passwordTextField.enabled = true
        submitButton.enabled = true
    }
}
